### ISSUE TITLE
Test suite fails with numpy.dtype issue

### ISSUE THREAD

**Issue by wjakob** [pybind/pybind11#355]
cc @aldanor 

Hi Ivan,

I just tried to build the test suite on a linux machine and it failed with a not-so-helpful error message ``NumPy: failed to create structured dtype``. Digging a bit further revealed that the following call to dtype via ``PyArray_DescrConverter`` is to blame:

```
>>> a ={'formats': [dtype('bool'), dtype('uint32'), dtype('float32')], 'itemsize': 1, 'names': ['x', 'y', 'z'], 'offsets': [0, 4, 8]}
>>> np.dtype(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: NumPy dtype descriptor requires 12 bytes, cannot override to smaller itemsize of 1
>>> np.version.version
'1.11.0'
```

What can be done to address this?

Thanks,
Wenzel

**Comment by wjakob** [comment]
Off-topic: The error messages from pytest when stuff goes wrong are really nice ;)

**Comment by wjakob** [comment]
Actually, was your code ever supposed to create a type like that? It does not make much sense to me. (This is on the Intel compiler, so it could be a weird compiler interaction.)

**Comment by aldanor** [comment]
On the off-topic (I meant to raise this as a separate issue but forgot) -- there's many places where we check if a Python-level error occurred and just do something like `pybind11_fail("oops")`. It would be a *lot* more helpful if in all of these cases we used the actual exception string which we can (try to) get from Python.

**Comment by aldanor** [comment]
Is this before or after #354?

The itemsize of 1 doesn't make much sense.

**Comment by wjakob** [comment]
before

**Comment by aldanor** [comment]
Not quite sure about this, would be nice to be able to reproduce. Could you check `sizeof(SimpleStruct)` from `example-numpy-dtypes.cpp`? Is this SimpleStruct that's failing, I guess (looking at the offsets, not the packed one?)

**Comment by aldanor** [comment]
You could also check that `args["itemsize"]` is indeed 12 and not 1 here: https://github.com/pybind/pybind11/blob/master/include/pybind11/numpy.h#L136

The itemsize argument is passed from here: https://github.com/pybind/pybind11/blob/master/include/pybind11/numpy.h#L412

There's literally the only two places I can think of where something could go wrong for some reason.

**Comment by wjakob** [comment]
Yep, that's the one that fails. I have ``sizeof(SimpleStruct) == 12``

**Comment by aldanor** [comment]
I think this should localize this to `dtype(list names, list formats, list offsets, size_t itemsize)` ctor then unless I'm missing something.

**Comment by wjakob** [comment]
Could it be related to some of the sorting stuff? 

When I add a the following line to 
```
std::cout << (std::string) args.str() << std::endl;
```
to ``dtype(list names, list formats, list offsets, size_t itemsize)``, I get

```
{'itemsize': 12, 'formats': [dtype('bool'), dtype('uint32'), dtype('float32')], 'offsets': [0, 4, 8], 'names': ['x', 'y', 'z']}
{'itemsize': 1, 'formats': [dtype('uint32'), dtype('float32'), dtype('bool')], 'offsets': [4, 8, 0], 'names': ['y', 'z', 'x']}
```

**Comment by wjakob** [comment]
I tried to replace all other usage of ``int_`` by ``pybind::int_`` but that didn't change anything (might be a good idea in any case though).

**Comment by wjakob** [comment]
fixed! (9a777a263dba76bbbe1cff173142d3e30b1363b3)

**Comment by wjakob** [comment]
I had missed some spots

**Comment by aldanor** [comment]
Was just writing about the same thing :) Nasty `int_`...

Does this fix it all? (most of those are cosmetic though as it should only be shadowed inside `detail`)

**Comment by wjakob** [comment]
Now I get these:

```
=========================================================================================== FAILURES ============================================================================================
_________________________________________________________________________________________ test_recarray _________________________________________________________________________________________

    @pytest.requires_numpy
    def test_recarray():
        from pybind11_tests import (create_rec_simple, create_rec_packed, create_rec_nested,
                                    print_rec_simple, print_rec_packed, print_rec_nested,
                                    create_rec_partial, create_rec_partial_nested)

        elements = [(False, 0, 0.0), (True, 1, 1.5), (False, 2, 3.0)]

        for func, dtype in [(create_rec_simple, simple_dtype), (create_rec_packed, packed_dtype)]:
>           arr = func(0)
E           RuntimeError: This should never happen: Invalid offset in buffer format string generation. Please report a bug to the Numpy developers.

test_numpy_dtypes.py:65: RuntimeError
_______________________________________________________________________________________ test_string_array _______________________________________________________________________________________

    @pytest.requires_numpy
    def test_string_array():
        from pybind11_tests import create_string_array, print_string_array

>       arr = create_string_array(True)
E       RuntimeError: This should never happen: Invalid offset in buffer format string generation. Please report a bug to the Numpy developers.

test_numpy_dtypes.py:138: RuntimeError
======================================================================== 2 failed, 58 passed, 10 skipped in 3.71 seconds ==========================================================
```

**Comment by wjakob** [comment]
Yes, the cast to size_t seems to have been problematic.

**Comment by aldanor** [comment]
This gets thrown in `_buffer_format_string` in `multiarray/buffer.c` when NumPy tries to generate the format string for a dtype. This, in turn, is only called from `_buffer_info_new` which is called from `_buffer_get_info` which is called from `array_getbuffer` -- so I think it fails upon `.request()` when NumPy realizes the dtype is *really* messed up.

Idk if it has anything to do with `strip_padding()`? (the `dtype(const buffer_info&)` ctor first creates a dtype using `numpy.core._internal._dtype_from_pep_3118` pure Python function, and then strips the padding and sorts the offsets).

I got to got now as it's a bit late, but will help tomorrow if it's not fixed by then :)

**Comment by wjakob** [comment]
I fear this is a beyond my level of NumPy expertise, but I'll try to collect some information that may help you.

I added a printf() which emits the name that is passed to _dtype_from_pep_3118 for the failing tests. That leads to the following output.

```
=========================================================================================== FAILURES ============================================================================================
_________________________________________________________________________________________ test_recarray _________________________________________________________________________________________

    @pytest.requires_numpy
    def test_recarray():
        from pybind11_tests import (create_rec_simple, create_rec_packed, create_rec_nested,
                                    print_rec_simple, print_rec_packed, print_rec_nested,
                                    create_rec_partial, create_rec_partial_nested)

        elements = [(False, 0, 0.0), (True, 1, 1.5), (False, 2, 3.0)]

        for func, dtype in [(create_rec_simple, simple_dtype), (create_rec_packed, packed_dtype)]:
>           arr = func(0)
E           RuntimeError: This should never happen: Invalid offset in buffer format string generation. Please report a bug to the Numpy developers.

test_numpy_dtypes.py:65: RuntimeError
------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------
T{=?:x:3x=I:y:=f:z:}
_______________________________________________________________________________________ test_string_array _______________________________________________________________________________________

    @pytest.requires_numpy
    def test_string_array():
        from pybind11_tests import create_string_array, print_string_array

>       arr = create_string_array(True)
E       RuntimeError: This should never happen: Invalid offset in buffer format string generation. Please report a bug to the Numpy developers.

test_numpy_dtypes.py:138: RuntimeError
------------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------------
T{=3s:a:=3s:b:}
________________________________________________________________________________________ test_vectorize _________________________________________________________________________________________

capture = <conftest.Capture object at 0x7f08396f6cc0>

    @pytest.requires_numpy
    def test_vectorize(capture):
        from pybind11_tests import vectorized_func, vectorized_func2, vectorized_func3

        assert np.isclose(vectorized_func3(np.array(3 + 7j)), [6 + 14j])

        for f in [vectorized_func, vectorized_func2]:
            with capture:
                assert np.isclose(f(1, 2, 3), 6)
            assert capture == "my_func(x:int=1, y:float=2, z:float=3)"
            with capture:
                assert np.isclose(f(np.array(1), np.array(2), 3), 6)
            assert capture == "my_func(x:int=1, y:float=2, z:float=3)"
            with capture:
                assert np.allclose(f(np.array([1, 3]), np.array([2, 4]), 3), [6, 36])
>           assert capture == """
                my_func(x:int=1, y:float=2, z:float=3)
                my_func(x:int=3, y:float=4, z:float=3)
            """
E           assert --- actual / +++ expected
E             - d
E               my_func(x:int=1, y:float=2, z:float=3)
E               my_func(x:int=3, y:float=4, z:float=3)

test_numpy_vectorize.py:22: AssertionError
======================================================================== 3 failed, 57 passed, 10 skipped in 3.24 seconds ========================================================================
```

**Comment by aldanor** [comment]
The buffer format strings look ok to me, need to dig around a bit deeper :)

I've requested an open source license for Intel Compiler, maybe they will give me one to debug this. In the meanwhile it would be pretty hard to get to the bottom of this without being able to reproduce..

**Comment by aldanor** [comment]
Ok, I got a trial of `icpc` for 31 days, can reproduce on OSX ✅

**Comment by aldanor** [comment]
@wjakob Fixed #358

**Comment by aldanor** [comment]
I've been able to run the full test suite successfully on the latest `icpc` on OS X, does it work for you?

**Comment by wjakob** [comment]
passes 👍
